From 11cb24076195a67b2a71e1874ffa2de0ec446c53 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:33 -0400
Subject: [scsi] qla2xxx: Log when device state is moved to failed state

Message-id: <1408391817-6389-46-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87947
O-Subject: [PATCH RHEL 7.1 e-stor 45/69] qla2xxx: Log when device state is moved to failed state.
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Giridhar Malavali <giridhar.malavali@qlogic.com>

scsi commit id 4fa0c6687f1b6ac43c562f03184bd88e88bb4ab3

Signed-off-by: Giridhar Malavali <giridhar.malavali@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_nx2.c b/drivers/scsi/qla2xxx/qla_nx2.c
index 08ee387..9bea074 100644
--- a/drivers/scsi/qla2xxx/qla_nx2.c
+++ b/drivers/scsi/qla2xxx/qla_nx2.c
@@ -1655,6 +1655,19 @@ qla8044_set_drv_active(struct scsi_qla_host *vha)
  qla8044_wr_direct(vha, QLA8044_CRB_DRV_ACTIVE_INDEX, drv_active);
 }
 
+static int
+qla8044_check_drv_active(struct scsi_qla_host *vha)
+{
+ uint32_t drv_active;
+ struct qla_hw_data *ha = vha->hw;
+
+ drv_active = qla8044_rd_direct(vha, QLA8044_CRB_DRV_ACTIVE_INDEX);
+ if (drv_active & (1 << ha->portnum))
+  return QLA_SUCCESS;
+ else
+  return QLA_TEST_FAILED;
+}
+
 static void
 qla8044_clear_idc_dontreset(struct scsi_qla_host *vha)
 {
@@ -1837,14 +1850,16 @@ qla8044_device_state_handler(struct scsi_qla_host *vha)
 
  while (1) {
   if (time_after_eq(jiffies, dev_init_timeout)) {
-   ql_log(ql_log_warn, vha, 0xb0cf,
-       "%s: Device Init Failed 0x%x = %s\n",
-       QLA2XXX_DRIVER_NAME, dev_state,
-       dev_state < MAX_STATES ?
-       qdev_state(dev_state) : "Unknown");
-
-   qla8044_wr_direct(vha, QLA8044_CRB_DEV_STATE_INDEX,
-       QLA8XXX_DEV_FAILED);
+   if (qla8044_check_drv_active(vha) == QLA_SUCCESS) {
+    ql_log(ql_log_warn, vha, 0xb0cf,
+        "%s: Device Init Failed 0x%x = %s\n",
+        QLA2XXX_DRIVER_NAME, dev_state,
+        dev_state < MAX_STATES ?
+        qdev_state(dev_state) : "Unknown");
+    qla8044_wr_direct(vha,
+        QLA8044_CRB_DEV_STATE_INDEX,
+        QLA8XXX_DEV_FAILED);
+   }
   }
 
   dev_state = qla8044_rd_direct(vha, QLA8044_CRB_DEV_STATE_INDEX);
-- 
1.7.1